lib/url_parser.js

```diff
@@ -4,9 +4,61 @@ var ReadPreference = require('./read_preference'),
   parser = require('url'),
   f = require('util').format,
   assign = require('./utils').assign,
-  Logger = require('mongodb-core').Logger;
+  Logger = require('mongodb-core').Logger,
+  dns = require('dns');
 
-module.exports = function(url, options) {
+module.exports = function(url, options, callback) {
+  var result = parser.parse(url, true);
+  if (result.protocol !== 'mongodb:') {
+    // TODO fix this
+    if (result.protocol !== 'mongodb+srv:') {
+      throw new Error('invalid schema, expected mongodb or mongodb+srv');
+    }
+  }
+
+  if (result.protocol === 'mongodb+srv:') {
+    if (result.port) {
+      return callback(new Error('Ports not accepted with mongodb+srv'));
+    }
+
+    const srvAddress = `_mongodb._tcp.${result.host}`;
+    dns.resolveSrv(srvAddress, function(err, addresses) {
+      if (err) return callback(err);
+
+      if (addresses.length === 0) {
+        return callback(new Error('No addresses found at host'));
+      }
+
+      dns.resolveTxt(result.host, function(err, records) {
+        if (err && err.code !== 'ENODATA') return callback(err);
+        if (err && err.code === 'ENODATA') records = null;
+
+        let connectionStrings = addresses.map(function(address, i) {
```

Comment:
for fun, this can be rewritten as such:
```
let connectionStrings = addresses.map((address, i) => (i === 0) ?
 `mongodb://${address.name}:${address.port}` : `${address.name}:${address.port}`);
```

Reply:
But also couldn't we do something like:
```
`mongodb://${addresses.map(address => `${address.name}:${address.port}`).join(',')}`
```

Reply:
that second option seems very obfuscating. I think we'd be better off using a temp variable.

Reply:
Actually, with the linter both of these get formatted in not great ways so gonna keep it as-is.